[build] Default to using NDK r23

jonpryor opened the pull request:
Context: 70272dbd9600dca4b161ecd2c0251ee68d2c3481
Context: 38b12367f959533ef69fe9858202a3097a4df903
Context: https://github.com/actions/virtual-environments/blob/9cf1ebd754807fba137c8ce9fa2140311609fdd6/images/win/Windows2022-Readme.md
Context: https://github.com/actions/virtual-environments/commit/2950cbfeab88a6a6202fa31d7371e574dbe2dc51

We've had an "inadvertent thinko" for the past couple of months:
in commit 70272dbd we updated the default `$(AndroidNdkDirectory)`
value to be `$(ANDROID_NDK_LATEST_HOME)`, if set.  *At the time*,
we think this was NDK r23.

In commit 38b12367, we updated `xaprepare` to install NDK r24.

The "thinko" is that commit 38b12367 was *irrelevant*, as commit
70272dbd was "in control".  The Windows **Prepare Solution** log
would contain e.g.:

	  Checking if android-ndk-r24-windows exists in C:\Android\android-sdk\ndk\23.2.8568313
	  Component 'android-ndk-r24-windows' requires Pkg.Revision to be '24.0.8215888', verifying
	      installed

It also *mis-detected* NDK r23 as r24 (?!).

It "worked" only by coincidence.

Around 2022-Jun-09, GitHub actions updated the default NDK installed
on their Windows images; previously, `$(ANDROID_NDK_LATEST_HOME)` was
NDK r23 or earlier (we're not sure); *now*, it is NDK r24.

This subtle change -- invisible!  The build environment changed! --
*introduced* unit test failures, of two patterns.

Pattern 1 is that `readelf` could not be found:

	System.ComponentModel.Win32Exception : An error occurred trying to start process
	'C:\Android\android-sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\readelf'
	with working directory 'C:\a_work\1\s\xamarin-android\bin\TestRelease\net6.0'.
	The system cannot find the file specified.
	   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.RunCommand(String executablePath, String arguments) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 550
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertSharedLibraryHasRequiredSymbols(String dsoPath, String readElfPath) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 495
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertValidEnvironmentSharedLibrary(String outputDirectoryRoot, String sdkDirectory, String ndkDirectory, String supportedAbis) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 489

	-or-
	System.ComponentModel.Win32Exception : The system cannot find the file specified
	   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.RunCommand(String executablePath, String arguments) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 551
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertSharedLibraryHasRequiredSymbols(String dsoPath, String readElfPath) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 495
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertValidEnvironmentSharedLibrary(String outputDirectoryRoot, String sdkDirectory, String ndkDirectory, String supportedAbis) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 489

This happened because `readelf` was *removed* in NDK r24.

Fix this pattern by updating `$(AndroidNdkDirectory)` to instead
default to `$(ANDROID_NDK_HOME)`, which is now NDK r23.
Additionally, update
`EnvironmentHelper.AssertValidEnvironmentSharedLibrary()` to also
look for `llvm-readelf`, the replacement for `readelf` in NDK r24+.

Pattern 2 is that the `aarch64-linux-android28-clang.cmd` & related
scripts on Windows don't support being run from a directory
containing spaces:

	CC="C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi19-clang.CMD"
	AS="C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\binutils\bin\arm-linux-androideabi-as.CMD"
	…
	[CC] "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi19-clang.CMD" ^
	  -c -D__ANDROID_API__=19 -DANDROID -o obj\Release\bundles\armeabi-v7a\temp.o ^
	  -I "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\include\arm-linux-androideabi" ^
	  -I "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\include" ^
	  obj\Release\bundles\armeabi-v7a\temp.c
	[cc stderr] 'C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK' is not recognized as an internal or external command,
	[cc stderr] operable program or batch file.

Fix this pattern by updating `NdkToolsWithClangNoBinutils` to
override the C and C++ compiler toolnames, a'la: 71ae5566.

jonathanpeppers reviewed (commented):
There might still be an issue using `mkbundle` on Windows:
```
(_BuildApkEmbed target) -> 
  C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028: System.IO.FileNotFoundException: Could not find file 'C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\obj\Release\bundles\armeabi-v7a\libmonodroid_bundle_app.so'. [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028: File name: 'C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\obj\Release\bundles\armeabi-v7a\libmonodroid_bundle_app.so' [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at System.IO.File.GetAttributes(String path) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Tools.Zip.WindowsPlatformServices.IsDirectory(ZipArchive archive, String path, Boolean& result) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Tools.Zip.PlatformServices.<>c__DisplayClass9_0.<IsDirectory>b__0(IPlatformServices services) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Tools.Zip.PlatformServices.CallServices(Func`2 code) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Tools.Zip.PlatformServices.IsDirectory(ZipArchive archive, String path) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Tools.Zip.ZipArchive.AddFile(String sourcePath, String archivePath, EntryPermissions permissions, CompressionMethod compressionMethod, Boolean overwriteExisting) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Android.Tasks.BuildApk.ExecuteWithAbi(String[] supportedAbis, String apkInputPath, String apkOutputPath, Boolean debug, Boolean compress, IDictionary`2 compressedAssembliesInfo, String assemblyStoreApkName) [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Xamarin.Android.Tasks.BuildApk.RunTask() [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
C:\a\_work\1\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2053,3): error XABLD7028:    at Microsoft.Android.Build.Tasks.AndroidTask.Execute() in C:\a\_work\1\s\external\xamarin-android-tools\src\Microsoft.Android.Build.BaseTasks\AndroidTask.cs:line 17 [C:\a\_work\1\a\TestRelease\06-14_02.21.17\temp\BuildMkBundleApplicationRelease\UnnamedProject.csproj]
```

grendello commented on src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/WithClangNoBinutils.cs:
```diff
@@ -14,6 +14,41 @@ public NdkToolsWithClangNoBinutils (string androidNdkPath, NdkVersion version, T
+				//
+				NdkToolNames[NdkToolKind.CompilerC] = "clang.exe";
+				NdkToolNames[NdkToolKind.CompilerCPlusPlus] = "clang++.exe";
+			}
```
This change won't be enough, it also needs [this code](https://github.com/xamarin/xamarin-android/blob/944f88a8b6f1c30eaf43f56cfbebf0f8cd7a547f/src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/WithClangWithPlatforms.cs#L49-L68) to mimic the .cmd wrapper script

grendello commented on Configuration.props:
```diff
@@ -93,7 +93,7 @@
     <AndroidSdkDirectory Condition=" '$(AndroidSdkDirectory)' == '' And Exists($(ANDROID_SDK_ROOT)) ">$(ANDROID_SDK_ROOT)</AndroidSdkDirectory>
     <AndroidSdkDirectory Condition=" '$(AndroidSdkDirectory)' == '' ">$(AndroidToolchainDirectory)\sdk</AndroidSdkDirectory>
-    <AndroidNdkDirectory Condition=" '$(AndroidNdkDirectory)' == '' And Exists($(ANDROID_NDK_LATEST_HOME)) ">$(ANDROID_NDK_LATEST_HOME)</AndroidNdkDirectory>
+    <AndroidNdkDirectory Condition=" '$(AndroidNdkDirectory)' == '' And Exists($(ANDROID_NDK_HOME)) ">$(ANDROID_NDK_HOME)</AndroidNdkDirectory>
```
I'm still confused why don't we just use the `~/android-toolchain/ndk` location that we fully control? If we want to use NDK that's already installed on the machine, to avoid downloading the NDK zip, perhaps `xaprepare` should be modified to check the system location for a matching version and copy the SDK to `~/android-toolchain/ndk`, if the version matches our requirements? That way we'd remove a known unknown of the system location.

grendello replied:
I know that https://github.com/xamarin/xamarin-android/commit/70272dbd9600dca4b161ecd2c0251ee68d2c3481 mentions the reason to be disk space, and that `xaprepare` overwrites the system location with "our" NDK if necessary, but that appeared to have failed in the past 3 months, since we thought we were using NDK r24 (@pjcollins checked that `xaprepare` **did** overwrite the system installation with our version) but we probably didn't? Or we used a mix of whatever NDK version was installed in the system location and what `xaprepare` overlaid on it (which would explain why we found `readelf` in there).

pjcollins replied:
Unfortunately this change will not force NDK 23 usage in the tests.  This will only change where we install NDK 24, we'll now use `$(ANDROID_NDK_HOME)` rather than `$(ANDROID_NDK_LATEST_HOME)`.

We are able to write to either location declared in these variables, and we replace the contents of the path entirely with the NDK required by xaprepare assuming it is not already installed. I looked at a handful of logs and this appears to be working as expected on macOS, Windows, and Linux. I agree that we do want full control over the NDK we use for build and test regardless of system image updates. Provisioning appears to be fine in this regard, the problem comes down to how our tests resolve the NDK.

The vast majority of MSBuild tests no longer set `/p:AndroidNdkDirectory=path`, so we are left with the default NDK lookup behavior in [xamarin-android-tools](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L175). This logic was never [updated to support NDK 24](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L14), and as a result it ignores the NDK 24 that we have installed, and will use a potentially arbitrary one:

    Verbose: Skipping NDK in 'C:\Android\android-sdk\ndk\24.0.8215888': version 24.0.8215888 is out of the accepted range (major version must be between 16 and 23
    Verbose: Best NDK selected: v23.2.8568313 in C:\Android\android-sdk\ndk\23.2.8568313

Before the recent VM image update we would fall back to v22:

    Message: DEBUG  Skipping NDK in 'C:\Program Files (x86)\Android\android-sdk\ndk\23.2.8568313': version 24.0.8215888 is out of the accepted range (major version must be between 16 and 23
    Message: DEBUG  Best NDK selected: v22.1.7171670 in C:\Program Files (x86)\Android\android-sdk\ndk\22.1.7171670

On Windows, another wrench is thrown into the works. The [AotTests fixture](https://github.com/xamarin/xamarin-android/blob/745214deb03343124f5a1610434f4746ddaf669c/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AotTests.cs#L31) will resolve an NDK [with this lookup logic](https://github.com/xamarin/xamarin-android/blob/7e8cc34989bfcb5bd58cb857013ba67273c00c79/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs#L38), and then copy it to a path with special characters to test against. Unfortunately this lookup prioritizes `$(ANDROID_NDK_HOME)` over `$(ANDROID_NDK_LATEST_HOME)`, introducing a second place where the tests can use an arbitrary NDK rather than the one we install.

I think we should continue to use `$(ANDROID_NDK_LATEST_HOME)` here for performance reasons, we should not always attempt to install NDK 24 as it is now present on our test machines. Instead we can change the [NDK lookup ordering](https://github.com/xamarin/xamarin-android/blob/7e8cc34989bfcb5bd58cb857013ba67273c00c79/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs#L38) that some of the tests use to prefer `$(ANDROID_NDK_LATEST_HOME)`. We will also want to bump the [`MaximumCompatibleNDKMajorVersion`](https://github.com/xamarin/xamarin-android-tools/blob/47832f1878e30446eebdafaf1f8b1ac5903bdca9/src/Xamarin.Android.Tools.AndroidSdk/Sdks/AndroidSdkBase.cs#L14) to allow us to use the expected NDK 24 version that we provision, assuming it doesn't have other major breaking changes.

Finally, we may want to make sure to set ` /p:AndroidNdkDirectory=$(ANDROID_NDK_LATEST_HOME)` for all tests that use llvm or mkbundle if this isn't being done already.

pjcollins replied:
To summarize, I think we should return to using `$(ANDROID_NDK_LATEST_HOME)` as the default NDK path to install into or use if the version matches, and limit the options our tests have to resolve the SDK/NDK to the following:

1) Test override variables
2) Variables used by CI (`$(ANDROID_NDK_LATEST_HOME)` and `$(ANDROID_NDK_LATEST_HOME)`)
3) Local install path (`~/android-toolchain`)

```diff
diff --git a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
index 54776ab15..31c8c4f19 100644
--- a/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
+++ b/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Android/AndroidSdkResolver.cs
@@ -22,10 +22,6 @@ namespace Xamarin.ProjectTools
 		public static string GetAndroidSdkPath ()
 		{
 			var sdkPath = Environment.GetEnvironmentVariable ("TEST_ANDROID_SDK_PATH");
-			if (String.IsNullOrEmpty (sdkPath))
-				sdkPath = Environment.GetEnvironmentVariable ("ANDROID_SDK_PATH");
-			if (String.IsNullOrEmpty (sdkPath))
-				sdkPath = GetPathFromRegistry ("AndroidSdkDirectory");
 			if (String.IsNullOrEmpty (sdkPath))
 				sdkPath = Environment.GetEnvironmentVariable ("ANDROID_SDK_ROOT");
 			if (String.IsNullOrEmpty (sdkPath))
@@ -38,10 +34,6 @@ namespace Xamarin.ProjectTools
 		public static string GetAndroidNdkPath ()
 		{
 			var ndkPath = Environment.GetEnvironmentVariable ("TEST_ANDROID_NDK_PATH");
-			if (String.IsNullOrEmpty (ndkPath))
-				ndkPath = Environment.GetEnvironmentVariable ("ANDROID_NDK_PATH");
-			if (String.IsNullOrEmpty (ndkPath))
-				ndkPath = GetPathFromRegistry ("AndroidNdkDirectory");
 			if (String.IsNullOrEmpty (ndkPath))
 				ndkPath = Environment.GetEnvironmentVariable ("ANDROID_NDK_LATEST_HOME");
 			if (String.IsNullOrEmpty (ndkPath))
```

jonpryor commented:
/azp run

jonpryor commented:
Proposed commit message:

```markdown
[build] Properly support using NDK r24

Context: 70272dbd9600dca4b161ecd2c0251ee68d2c3481
Context: 38b12367f959533ef69fe9858202a3097a4df903
Context: https://github.com/actions/virtual-environments/blob/9cf1ebd754807fba137c8ce9fa2140311609fdd6/images/win/Windows2022-Readme.md
Context: https://github.com/actions/virtual-environments/commit/2950cbfeab88a6a6202fa31d7371e574dbe2dc51

Changes: http://github.com/xamarin/xamarin-android-tools/compare/20f611202bef0fc7c1659366dd38865eb119dde5...ec346d07cf3ac7fc74d08eae4f19043b51485724

  * xamarin/xamarin-android-tools@ec346d0: [Xamarin.Android.Tools.AndroidSdk] Permit NDK r24 (#171)
  * xamarin/xamarin-android-tools@47832f1: [Xamarin.Android.Tools.AndroidSdk] AndroidSdkInfo validation locator (#170)

We've had an "inadvertent behavior" for the past couple of months:
in commit 70272dbd we updated the default `$(AndroidNdkDirectory)`
value to be `$(ANDROID_NDK_LATEST_HOME)`, if set.  *At the time*,
we think this was NDK r23.

In commit 38b12367, we updated `xaprepare` to install NDK r24.

This appears to have caused `xaprepare` to update the *system*
`$(ANDROID_NDK_LATEST_HOME)` location, not a xamarin-android-specific
NDK installation.  The Windows **Prepare Solution** log would
contain e.g.:

	  Checking if android-ndk-r24-windows exists in C:\Android\android-sdk\ndk\23.2.8568313
	  Component 'android-ndk-r24-windows' requires Pkg.Revision to be '24.0.8215888', verifying
	      installed

This appears to imply that `C:\Android\android-sdk\ndk\23.2.8568313`
was updated to have NDK 24.0.8215888.

Meanwhile, *unit tests* were using `$(ANDROID_SDK_PATH)` and
`$(ANDROID_NDK_PATH)`, which had been using NDK *r21*.

Around 2022-Jun-09, GitHub actions updated the default NDK installed
on their Windows images; previously, `$(ANDROID_NDK_LATEST_HOME)` was
NDK r23 or earlier (we're not sure); *now*, it is NDK r24.
`$(ANDROID_NDK_PATH)` had been NDK r21; now it is NDK r23.

This subtle change -- invisible!  The build environment changed! --
*introduced* unit test failures, of two patterns.

Scenario 1 is that `readelf` could not be found:

	System.ComponentModel.Win32Exception : An error occurred trying to start process
	'C:\Android\android-sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\readelf'
	with working directory 'C:\a_work\1\s\xamarin-android\bin\TestRelease\net6.0'.
	The system cannot find the file specified.
	   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.RunCommand(String executablePath, String arguments) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 550
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertSharedLibraryHasRequiredSymbols(String dsoPath, String readElfPath) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 495
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertValidEnvironmentSharedLibrary(String outputDirectoryRoot, String sdkDirectory, String ndkDirectory, String supportedAbis) in C:\a\_work\1\s\src\Xamarin.Android.Build.Tasks\Tests\Xamarin.Android.Build.Tests\Utilities\EnvironmentHelper.cs:line 489

	-or-
	System.ComponentModel.Win32Exception : The system cannot find the file specified
	   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.RunCommand(String executablePath, String arguments) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 551
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertSharedLibraryHasRequiredSymbols(String dsoPath, String readElfPath) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 495
	   at Xamarin.Android.Build.Tests.EnvironmentHelper.AssertValidEnvironmentSharedLibrary(String outputDirectoryRoot, String sdkDirectory, String ndkDirectory, String supportedAbis) in /Users/builder/azdo/_work/1/s/xamarin-android/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:line 489

This happened because `readelf` was *removed* in NDK r24.

Fix this scenario updating
`EnvironmentHelper.AssertValidEnvironmentSharedLibrary()` to also
look for `llvm-readelf`, the replacement for `readelf` in NDK r24+.

Scenario 2 is that the `aarch64-linux-android28-clang.cmd` & related
scripts on Windows in NDK r23+ don't support being run from a
directory containing spaces:

	CC="C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi19-clang.CMD"
	AS="C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android\binutils\bin\arm-linux-androideabi-as.CMD"
	…
	[CC] "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi19-clang.CMD" ^
	  -c -D__ANDROID_API__=19 -DANDROID -o obj\Release\bundles\armeabi-v7a\temp.o ^
	  -I "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\include\arm-linux-androideabi" ^
	  -I "C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK Ümläüts\ndk\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\include" ^
	  obj\Release\bundles\armeabi-v7a\temp.c
	[cc stderr] 'C:\a\_work\1\a\TestRelease\06-13_18.07.09\temp\SDK' is not recognized as an internal or external command,
	[cc stderr] operable program or batch file.

Fix this scenario by updating `NdkToolsWithClangNoBinutils` to
override the C and C++ compiler toolnames, a'la commit 71ae5566.

To help make investigating
`AndroidDependenciesTests.InstallAndroidDependenciesTest()` tests
failures easier, send the `InstallAndroidDependencies` target
execution output to an `install-deps.log` file.  Previously, the
subsequent `Build` target would *overwrite* the output of the
`InstallAndroidDependencies` output, as they both shared `build.log`.

Finally, update `AndroidSdkResolver.GetAndroidSdkPath()` and
`AndroidSdkResolver.GetAndroidNdkPath()` to *stop* looking at the
`$(ANDROID_SDK_PATH)`, `$(ANDROID_SDK_ROOT)`, `$(ANDROID_NDK_PATH)`,
and `$(ANDROID_NDK_LATEST_HOME)` environment variables.  Instead,
unit tests should use the `$(TEST_ANDROID_SDK_PATH)` environment
variable, which is managed by the unit test infrastructure.
```

jonpryor commented:
`BuildAMassiveApp()` failure is:

```
error XBD001: Download failed. Please download https://dl.google.com/dl/android/maven2/com/google/android/…/play-services-base/17.6.0/play-services-base-17.6.0.aar to a file called C:\Users\cloudtest\AppData\Local\XamarinBuildDownloadCache\playservicesbase-17.6.0.aar
```

Ignoring…
